### PR TITLE
Adding HPO unit test

### PR DESCRIPTION
**Pull request by gautamkmr**
```
786, 793, 800-809, 814-828, 833-839, 874, 877-878
/Users/gauta/pipelines/components/aws/sagemaker/deploy/src/deploy.py                                    44     12    73%   56-69, 73
/Users/gauta/pipelines/components/aws/sagemaker/ground_truth/src/ground_truth.py                        53     15    72%   55-71, 75
/Users/gauta/pipelines/components/aws/sagemaker/hyperparameter_tuning/src/hyperparameter_tuning.py      67      1    99%   95
/Users/gauta/pipelines/components/aws/sagemaker/model/src/create_model.py                               31     10    68%   38-49, 53
/Users/gauta/pipelines/components/aws/sagemaker/train/src/train.py                                      52      1    98%   82
/Users/gauta/pipelines/components/aws/sagemaker/workteam/src/workteam.py                                26     10    62%   33-44, 48
----------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                  892    259    71%

```

### CODE REVIEW

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 1]
```diff
@@ -60,4 +64,286 @@ def test_empty_string(self):
...
+
+    # Ensure all of the optional arguments have reasonable default values
+    self.assertFalse(response['TrainingJobDefinition']['EnableManagedSpotTraining'])
+    self.assertDictEqual(response['TrainingJobDefinition']['StaticHyperParameters'], {})
+    self.assertNotIn('VpcConfig', response['TrainingJobDefinition'])
+    self.assertNotIn('MetricDefinitions', response['TrainingJobDefinition'])
+    self.assertEqual(response['Tags'], [])
+    self.assertEqual(response['TrainingJobDefinition']['AlgorithmSpecification']['TrainingInputMode'], 'File')
+    self.assertEqual(response['TrainingJobDefinition']['OutputDataConfig']['S3OutputPath'], 'test-output-location')
+
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:191]: Can we add some more good defaults into this test? This is just sanity checking that the default arguments produce proper hyperparameters and solid default outputs/inputs.

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 2]
```diff
@@ -60,4 +64,286 @@ def test_empty_string(self):
...
+    required_vpc_args = self.parser.parse_args(required_args + ['--vpc_security_group_ids', 'sg1,sg2', '--vpc_subnets', 'subnet1,subnet2'])
+    response = _utils.create_hyperparameter_tuning_job_request(vars(required_vpc_args))
+    
+    self.assertIn('TrainingJobDefinition', response)
+    self.assertIn('VpcConfig', response['TrainingJobDefinition'])
+    self.assertIn('sg1', response['TrainingJobDefinition']['VpcConfig']['SecurityGroupIds'])
+    self.assertIn('sg2', response['TrainingJobDefinition']['VpcConfig']['SecurityGroupIds'])
+    self.assertIn('subnet1', response['TrainingJobDefinition']['VpcConfig']['Subnets'])
+    self.assertIn('subnet2', response['TrainingJobDefinition']['VpcConfig']['Subnets'])
+
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:344]: Can we add a test to ensure we have good support for each type of "ParameterRanges": Categorical, Continuous and Integer?

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 3]
```diff
@@ -60,4 +64,286 @@ def test_empty_string(self):
...
+      call('/tmp/best_job_name.txt', 'w'),
+      call('/tmp/best_hyperparameters.txt', 'w'),
+      call('/tmp/model_artifact_url.txt', 'w'),
+      call('/tmp/training_image.txt', 'w')
+    ], any_order=True)
+
+    file_open().write.assert_has_calls([
+      call('job-name'),
+      call('best_job'),
+      call('"best_hyperparameters"'),
```
[the remark this answers is not in the document]
Comment (gautamkmr) [reply components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:99]: not sure but this is what we see 
```
E               Actual: [call('job-name'),
E                call('best_job'),
E                call('"best_hyperparameters"'),
E                call('model-artifacts'),
E                call('training-image')]
```
So I am using this from output.

[the remark this answers is not in the document]
Comment (RedbackThomson) [reply components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:99]: Alright. I'm guessing it's doing a `json.dumps()` of the hyperparam dictionary, then.

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py (hunk 1 above)
Comment (gautamkmr) [reply components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:191]: I think its getting tested in `test_create_hyperparameter_tuning_job` with all default value that`create_hyper_parameter_tuning_job` uses.

Comment (RedbackThomson) [reply components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:191]: Sure yeah that sounds fine.

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def test_spot_local_path(self):
     self.assertEqual(response['TrainingJobDefinition']['CheckpointConfig']['S3Uri'], 's3://fake-uri/')
     self.assertEqual(response['TrainingJobDefinition']['CheckpointConfig']['LocalPath'], 'local-path')
 
-  def test_empty_string(self):
+  def check_empty_string(self):
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:62]: I still don't feel this really describes what the test is trying to achieve. I think this test name is used in every file, now that I look into it. So perhaps just for consistency we should just revert it back. Sorry about that.

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 5]
```diff
@@ -177,6 +177,37 @@ def test_get_image_from_algorithm_job(self):
...
+    # specifying both params 
+    good_args = self.parser.parse_args(required_args + ['--warm_start_type', 'TransferLearning'] + ['--parent_hpo_jobs', 'A,B,C'])
+    response = _utils.create_hyperparameter_tuning_job_request(vars(good_args))
+    self.assertIn('WarmStartConfig', response)
+    self.assertIn('ParentHyperParameterTuningJobs', response['WarmStartConfig'])
+    self.assertIn('WarmStartType', response['WarmStartConfig'])
+    self.assertEqual(response['WarmStartConfig']['ParentHyperParameterTuningJobs'][0]['HyperParameterTuningJobName'], 'A')
+    self.assertEqual(response['WarmStartConfig']['ParentHyperParameterTuningJobs'][1]['HyperParameterTuningJobName'], 'B')
+    self.assertEqual(response['WarmStartConfig']['ParentHyperParameterTuningJobs'][2]['HyperParameterTuningJobName'], 'C')
+    self.assertEqual(response['WarmStartConfig']['WarmStartType'], 'TransferLearning')
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:198]: ++ Great tests

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 6]
```diff
@@ -177,6 +177,37 @@ def test_get_image_from_algorithm_job(self):
...
+ 
+  def test_either_warm_start_or_parents_args(self):
+    # It will generate an exception if either warm_start_type or parent hpo jobs is being passed
+    good_args = self.parser.parse_args(required_args + ['--warm_start_type', 'TransferLearning'])
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:202]: nit: variable name (these are bad args)

File: components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py [hunk 7]
```diff
@@ -177,6 +177,37 @@ def test_get_image_from_algorithm_job(self):
...
+    with self.assertRaises(Exception):
+      _utils.create_hyperparameter_tuning_job_request(vars(good_args))
+
+    good_args = self.parser.parse_args(required_args + ['--parent_hpo_jobs', 'A,B,C'])
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py:206]: nit: variable name

**Pull request merged**